README.md

[![Build Status](https://travis-ci.org/loolaz/redis-mutex-semaphore.svg?branch=master)](https://travis-ci.org/loolaz/redis-mutex-semaphore)
[![Test Coverage](https://codeclimate.com/github/loolaz/redis-mutex-semaphore/badges/coverage.svg)](https://codeclimate.com/github/loolaz/redis-mutex-semaphore/coverage)
[![Code Climate](https://codeclimate.com/github/loolaz/redis-mutex-semaphore/badges/gpa.svg)](https://codeclimate.com/github/loolaz/redis-mutex-semaphore)
[![npm version](https://badge.fury.io/js/redis-mutex-semaphore.svg)](http://badge.fury.io/js/redis-mutex-semaphore)

# redis-mutex-semaphore
This is a mutex and semaphore library which is simply implemented by using basic redis commands.

```sh
npm install redis-mutex-semaphore
```

**Note:** The minimum Redis version is 2.6.0+.

## Constructing Instances

**createSemaphoreClient(key, count, [function callback(err, semaphoreClient){}])**

**createSemaphoreClient(key, count).then(function(semaphoreClient){})**
- semaphoreClient : semaphoreClient object for success / null for fail

**createMutexClient(key, ttl, [function callback(err, mutexClient){}])**

**createMutexClient(key, ttl).then(function(mutexClient){})**

- mutexClient : mutexClient object for success / null for fail

Whenever createXXXClient method is called, the redis key that you have passed through is also reset. So please be careful that you do not lose any context of these objects by mistakenly calling this method with the same key again in other places(other codes in the same process/other processes/other machines) while using the mutex/semaphore with the key. If you don't want to remove a key in redis if there exists, do not call createXXXClient method but call getXXXClient immediately(from v0.3.6).

```js
var redisSharedObject = require('redis-mutex-semaphore');
var factory = redisSharedObject({
  		host: '127.0.0.1',
  		port: 6379,
  		db: 1
  	});
  	
  	// or you can reuse existing redis connection
  
var factory = redisSharedObject(redisClient);	

factory.createSemaphoreClient('Key', 3 /* initial semaphore count */, function(err, semaphoreClient){}); // returns promise if callback is omitted
factory.createMutexClient('Key', 10 /* ttl : second */, function(err, mutexClient){});  // returns promise if callback is omitted

var semaphoreClient = factory.getSemaphoreClient('Key'),
    mutexClient = factory.getMutexClient('Key');

factory.end(); 
```

## Method Usage

### 1. Using CallBack

Same as creating xxxClient methods, you can call methods with callbacks.

#### 1.1. get & release method

**semaphoreClient.get(function callback(err, result){})**
 - result : true for success / false for fail to accquire
 - if key doesn't exist, ENOTFOUNDKEY error code is returned
 
**semaphoreClient.rel(function callback(err, result){})**
 - result : the number of remained semaphore for success

```js
semaphoreClient.get(function(err, result){
  if(result){
    // doing something with semaphore
    semaphoreClient.rel(function(err, result){
  
    });
  }
});
```

**mutexClient.get(function callback(err, result){})**
 - result : mutexid(uuid v4) for success / null for failed to lock
 
**mutexClient.rel(mutexID, function callback(err, result){})**
 - result : true for sucess / false for fail to delete key
 - if incorrect mutex id is passed, ENOACCESS error code is returned

```js
mutexClient.get(function(err, mutexID){
  if(mutexID){
    //doing something with lock
    mutexClient.rel(mutexID, function(err, result){
  
    });
  }
});
```

#### 1.2. wait/observe method

Both waitingFor/waitingForWithPriority and observe methods are blocked for a shared object to be released.
The difference between them is that waitingFor is blocked until getting a lock or timedout. However observing is blocked, and just returns when the observed object is released. 

The order of dispatching waiting clients is determined by considering their waiting priorities. FIFO(first in, first out) policy is also applied to the semaphore/mutexClient if it shares a redis connection. Other observing clients and event listeners are not affected by this scheduling policy.

**semaphoreClient/mutexClient.waitingFor(timeout, function callback(err, result){})**
 - result(semaphore) : true for success / false for fail to accquire
 - result(mutex) : mutex id for success / null for fail to lock
 - err(semaphore/mutex) : ETIMEDOUT error or other errors returned

**semaphoreClient/mutexClient.waitingForWithPriority(priority, timeout, function(err, result){})**
 - same as waitingFor except priority
 - client waits for semaphore/mutex with its own priority and will be scheduled according to it.
 - err(semaphore/mutex) : ETIMEDOUT error or other errors returned

priority argument takes follows:
 - redisSharedObject.priority.HIGH : immediate execution(default)
 - redisSharedObject.priority.NORMAL : random delay between 15~20ms
 - redisSharedObject.priority.LOW : random delay between 40~45ms

You can change default priority with **semaphoreClient/mutexClient.setDefaultPriority(priority)** method.
 
**semaphoreClient/mutexClient.observing(timeout, function callback(err, result){})**
 - result(semaphore/mutex) : true for success / false when ETIMEDOUT or other errors 
 - err(semaphore/mutex) : ETIMEDOUT error or other errors returned
 
```js
var redisSharedObject = require('redis-mutex-semaphore');

...

semaphoreClient/mutexClient.waitingFor(10 /* timeout : second */, function(err, result){
  if(result)
    // doing something with semaphore/lock
});

semaphoreClient/mutexClient.waitingForWithPriority(redisSharedObject.priority.NORMAL, 10 /* timeout : second */, function(err, result){
  if(result)
    // doing something with semaphore/lock
});

semaphoreClient/mutexClient.observing(10 /* timeout : second */, function(err, result){
  if(result)
    // doing something or try to accquire/lock
});

```

### 2. Using Promise

If callback is omitted, you can use it with promise.

#### 2.1. get & release method 

**semaphoreClient.get().then(function(result){})**
 - same as callback version
 
**semaphoreClient.rel().then(function(result){})**
 - same as callback version
 
```js
semaphoreClient.get().then(function(result){
  if(result){ 
    // doing something with semaphore
    return Promise.resolve(result);
  }
  else
    // doing others or return Promise.reject()
}).then(function(result){
  if(result)
    return semaphoreClient.rel();
}).catch(function(e){

});
```

**mutexClient.get().then(function(mutexID){})**
 - same as callback version
 
**mutexClient.rel(mutexID).then(function(result){})**
 - same as callback version
 
```js
mutexClient.get().then(function(mutexID){
  if(mutexID){
    // doing something with lock
    return Promise.resolve(mutexID);
  }
  else
    // doing others or return Promise.reject()
}).then(function(mutexID){
  return mutexClient.rel(mutexID);
}).catch(function(e){

});
```

#### 2.2. wait/observe method

**semaphoreClient/mutexClient.waitingFor(timeout).then(function(result){}).catch(function(err){})**
 - same as callback version

**semaphoreClient/mutexClient.waitingForWithPriority(priority, timeout).then(function(result){}).catch(function(err){})**
 - same as callback version

priority argument takes follows:
 - same as callback version

You can change default priority with **semaphoreClient/mutexClient.setDefaultPriority(priority)** method.

**semaphoreClient/mutexClient.observing(timeout).then(function(result){}).catch(function(err){})**
 - same as callback version

```js
var redisSharedObject = require('redis-mutex-semaphore');

...

semaphoreClient/mutexClient.waitingFor(10 /* timeout : second */).then(function(result){
  if(result)
    // doing something with semaphore/lock
}).catch(function(e){
  // e could be timed out error or others
});

semaphoreClient/mutexClient.waitingForWithPriority(redisSharedObject.priority.NORMAL, 10 /* timeout : second */).then(function(result){
  if(result)
    // doing something with semaphore/lock
}).catch(function(e){
  // e could be timed out error or others
});

semaphoreClient/mutexClient.observing(10 /* timeout : second */).then(function(result){
  if(result)
    // doing something or try to accquire/lock
}).catch(function(e){
  // e could be timed out error or others;
});
```

### 3. Others

**Checking status of queue**

This method returns value of mutex lock or semaphore's count and the sum of waiting/observing clients with a short delay.
[timeout] argument is optional, and unit is millisecond.

**Note:** It just broadcasts request for status checking to every subscribing clients and waits for their responses until timeout is reached. So if clients are spread out across other processes or network, the timeout value should be big enough to get all of the responses. The default value is 300 - this value is adjusted empirically by testing on the environment that the redis server is not far away from the service using this module geographically.

```js
semaphoreClient/mutexClient.getStaus([timeout, ] function callback(err, result){}) // callback
semaphoreClient/mutexClient.getStaus([timeout]).then(function(result){}) // promise

/* 
sample return value
{ 
  observing: 4, // the sum of observing clients
  waiting: 3, // the sum of waiting clients
  value: 0, // mutex id(string) or semaphore count(integer)
  observingLocal: 1, // the number of observing clients in current local instance
  waitingLocal: 0 // the number of waiting clients in current local instance
}
*/

```

Result object contains value, the number of waiting, the number of observing
- for mutex, value is mutex's id
- for semaphore, value is current semaphore count

**Reset semaphore/mutex**

```js
semaphoreClient/mutexClient.reset(function callback(err, result){}) // callback
semaphoreClient/mutexClient.reset().then(function(result){}) // promise

semaphoreClient/mutexClient.resetWithPublish(function callback(err, result){}) // callback
semaphoreClient/mutexClient.resetWithPublish().then(function(result){}) // promise
```

**Extend mutex timeout**

```js
mutexClient.extend(mutexID, more, function callback(err, result){}) // callback
mutexClient.extend(mutexID, more).then(function(result){}) // promise
```
 - result : true for success, otherwise false.


```js
factory.createMutexClient('key', 10); // timeout is set to 10 sec.

...

mutexClient.get(function(err, mutexID){  // mutex's is valid for 10 sec.
  if(mutexID){
    // doing something with mutex for N sec.
    mutexClient.extend(mutexID, 2, function(err, result){ // at this point, mutex is valid for 10 - N sec.
      if(result){ // succeeded to extend  
        // now mutex is valid for (10 - N + 2) sec.
        // doing something more...
      }
      else{ // faild to extend
        // mutex is already expired or failed to extend for other reasons
      }
    });
  }
  else {
    // failed to get mutex
  }
});
```

## Events

You can register listeners to the following events.

```js
semaphoreClient/mutexClient.on(event_name, function(arg){
  // doing something
});
```

| Name                | Description   | 
|---------------------|---------------|
| semaphore_acquired  | fired with the number of remained semaphore when semaphore is acquired. | 
| semaphore_released  | fired with the number of remained semaphore when semaphore is released.   | 
| mutex_locked        | fired with mutex id when mutex is locked. | 
| mutex_unlocked      | fired when getting a lock. it always returns 1.   | 
| mutex_expired       | fired when releasing a lock. it always returns 1. this event is sent to every clients.      |
| expired             | fired when mutex is expired. it always returns 1. this event is only sent to the client which has the lock.    |

## Run Tests

```
git clone git@github.com:loolaz/redis-mutex-semaphore.git
npm install
npm test
```

## Changes
see the [CHANGELOG](https://github.com/loolaz/redis-mutex-semaphore/blob/master/CHANGELOG.md)



CHANGELOG.md
# CHANGELOG

## 0.3.6

* getXXXClient method creates an object when there is no available object

## 0.3.4

* Fixed the bug that constructor's callback was not returned

## 0.3.3

* Minor bug fix

## 0.3.2

* Changed configuration value for(delay time) waitingForWithPriority
* Modified README instruction

## 0.3.1

* Fixed potential buts in some test cases
* Modified incorrect README instruction
* Modified ci configuration

## 0.3.0

* Removed setNewConnectionPerTransaction method(You don't need to call this method any more in any case)
* Changed minimum Redis version(2.2.0 => 2.6.0)
* Fixed potential bugs in some test cases
* Code Refactoring

## 0.2.12

* Added Mutex extend method(Redis 2.6.0+)
* Added changelog


## 0.2.10

* Changed default timeout value for getStatus method to 300ms
* Changed waiting priority setting
* Fixed minor bugs


test/index.js
'use strict';

var async = require('async'),
	redisSharedObject1,
	redisSharedObject2,
	muid,
	testSemaphoreKey = 'testObjectSem',
	testMutexKey1 = 'testObjectMutex1',
	testMutexKey2 = 'testObjectMutex2',
	options = {
		host: '127.0.0.1',
		port: 6379,
		db: 1
	};

describe('redis shared object test', function(){
	var RedisSharedObject = require('../lib');

	it('initialize', function(done){
		redisSharedObject1 = RedisSharedObject(options);
		redisSharedObject2 = RedisSharedObject(options);

		redisSharedObject1.createSemaphoreClient(testSemaphoreKey, 3);
		redisSharedObject1.createMutexClient(testMutexKey1, 10);

		redisSharedObject2.createSemaphoreClient(testSemaphoreKey, 3);
		redisSharedObject2.createMutexClient(testMutexKey1, 10);
		setTimeout( function(){ 
			done(); 
		}, 3000);
	});

	it('get three semaphore sequentially', function(done){
		var redisSemaphore1 = redisSharedObject1.getSemaphoreClient(testSemaphoreKey),
			redisSemaphore2 = redisSharedObject2.getSemaphoreClient(testSemaphoreKey);
		redisSemaphore1.get(function(err, sem){
			if(err)
				console.log(err);
			else
				console.log('first semaphore : ' + sem);

			redisSemaphore1.get(function(err, sem){
				if(err)
					console.log(err);
				else
					console.log('second semaphore : ' + sem);

				redisSemaphore2.get(function(err, sem){
					if(err)
						console.log(err);
					else
						console.log('third semaphore : ' + sem);	

					async.parallel([
						function(callback){
							redisSemaphore1.getStatus(function(err, result){
								expect(result.value).toEqual(0);
								callback(null, true);
							});
						},
						function(callback){
							redisSemaphore2.getStatus(function(err, result){
								expect(result.value).toEqual(0);
								callback(null, true);
							});
						}
						],
						function(err, result){
							done();
						}
					);
				});
			});
		});	
	}, 6000);

	it('release three semaphore, but six will compete to get them..', function(done){
		var redisSemaphore1 = redisSharedObject1.getSemaphoreClient(testSemaphoreKey),
			redisSemaphore2 = redisSharedObject2.getSemaphoreClient(testSemaphoreKey);		
		async.parallel([
			function(callback){
				redisSemaphore1.rel(function(err, result){
					if(err)
						console.log(err);
					else
						console.log('release result : ' + result);
					callback(err, result);
				});					
			},
			function(callback){
				redisSemaphore1.rel(function(err, result){
					if(err)
						console.log(err);
					else
						console.log('release result : ' + result);
					callback(err, result);
				});					
			},
			function(callback){
				redisSemaphore2.rel(function(err, result){
					if(err)
						console.log(err);
					else
						console.log('release result : ' + result);
					callback(err, result);
				});					
			},				
			function(callback){
				redisSemaphore1.waitingFor(5, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(1) : ' + result);
				});	
				callback(null, true);				
			},
			function(callback){
				redisSemaphore1.waitingFor(6, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(2) : ' + result);
				});	
				callback(null, true);				
			},
			function(callback){
				redisSemaphore1.waitingFor(6, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(3) : ' + result);
				});
				callback(null, true);					
			},
			function(callback){
				redisSemaphore2.waitingFor(6, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(4) : ' + result);
				});		
				callback(null, true);			
			},
			function(callback){
				redisSemaphore2.waitingFor(6, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(5) : ' + result);					
				});			
				callback(null, true);		
			},
			function(callback){
				redisSemaphore2.waitingFor(6, function(err, result){
					if(err)
						console.log(err);
					else
						console.log('tried and finally got one(6) : ' + result);					
				});		
				callback(null, true);			
			},
			function(callback){
				redisSemaphore2.observing(6, function(err, result){
					console.log('just waited, and woke up : ' + result);
					callback(err, result);
				});					
			}
			],function(err, result){
				setTimeout( function(){ 
					var firstCount = 0,
						firstWaiting = 0,
						firstObserving = 0,
						secondCount = 0,
						secondWaiting = 0,
						secondObserving = 0;
					async.parallel([
						function(callback){
							redisSemaphore1.getStatus(function(err, result){
								firstCount += result.value;
								firstWaiting += result.waiting;
								firstObserving += result.observing;
								callback(null, true);
							});
						},
						function(callback){
							redisSemaphore2.getStatus(function(err, result){
								secondCount += result.value;
								secondWaiting += result.waiting;
								secondObserving += result.observing;
								callback(null, true);
							});
						}
						],
						function(err, result){
							expect(firstCount).toEqual(0);
							expect(firstCount).toEqual(secondCount);
							expect(firstWaiting+secondWaiting).toEqual(3);
							expect(firstObserving+secondObserving).toEqual(0);
							done();
						}
					);
				}, 3000);
			}
		);

	}, 20000);

	it('release three semaphores', function(done){
		var redisSemaphore1 = redisSharedObject1.getSemaphoreClient(testSemaphoreKey),
			redisSemaphore2 = redisSharedObject2.getSemaphoreClient(testSemaphoreKey);		
		async.parallel([
			function(callback){
				redisSemaphore1.rel(function(err, sem){
					if(err)
						console.log(err);
					else
						console.log('relSem : ' + sem);
					callback(err, sem);
				});					
			},
			function(callback){
				redisSemaphore2.rel(function(err, sem){
					if(err)
						console.log(err);
					else
						console.log('relSem : ' + sem);
					callback(err, sem);
				});					
			},
			function(callback){
				redisSemaphore1.rel(function(err, sem){
					if(err)
						console.log(err);
					else
						console.log('relSem : ' + sem);
					callback(err, sem);
				});					
			}
			],function(err, result){
				setTimeout( function(){ 
					var firstCount = 0,
						firstWaiting = 0,
						firstObserving = 0,
						secondCount = 0,
						secondWaiting = 0,
						secondObserving = 0;
					async.parallel([
						function(callback){
							redisSemaphore1.getStatus(function(err, result){
								firstCount += result.value;
								firstWaiting += result.waiting;
								firstObserving += result.observing;
								callback(null, true);
							});
						},
						function(callback){
							redisSemaphore2.getStatus(function(err, result){
								secondCount += result.value;
								secondWaiting += result.waiting;
								secondObserving += result.observing;
								callback(null, true);
							});
						}
						],
						function(err, result){
							expect(firstCount).toEqual(secondCount);
							expect(firstCount).toEqual(0);
							expect(firstWaiting+secondWaiting).toEqual(0);
							expect(firstObserving+secondObserving).toEqual(0);
							done();
						}
					);
				}, 3000);
			}
		);

	}, 20000);

	it('get mutex and another waits for being released', function(done){
		var redisMutex1 = redisSharedObject1.getMutexClient(testMutexKey1),
			redisMutex2 = redisSharedObject2.getMutexClient(testMutexKey1);
		async.parallel([
			function(callback){
				redisMutex1.get(function(err, result){
					if(err)
						console.log(err);
					else{
						console.log('got mutex(1) : ' + result);
						if(result)
							muid = result;
					}
					callback(err, result);
				});					
			},
			function(callback){
				redisMutex2.get(function(err, result){
					if(err)
						console.log(err);
					else{
						console.log('got mutex(2) : ' + result);
						if(result)
							muid = result;
					}
					callback(err, result);
				});					
			},
			function(callback){
				redisMutex1.observing(8, function(err, released){
					if(err)
						console.log('err while observing : ' + err);
					else{
						console.log('sucessfully observed releasing : ' + released);
					}
					
				});
				callback(null, true);					
			},
			function(callback){
				redisMutex2.waitingFor(1, function(err, result){
					if(err)
						console.log('err while waiting(3) : ' + err);
					else{
						console.log('got mutex(3) : ' + result);
						if(result)
							muid = result;
					}
					expect(err.message).toEqual('timedout');
				});
				callback(null, true);					
			},
			function(callback){
				redisMutex1.waitingFor(8, function(err, result){
					if(err)
						console.log('err while waiting(4) : ' + err);
					else{
						console.log('got mutex(4) : ' + result);
						if(result)
							muid = result;
					}
					
				});
				callback(null, true);					
			}
			],function(err, result){
				setTimeout( function(){ 
					var actualmuid,
						firstWaiting = 0,
						firstObserving = 0,
						secondWaiting = 0,
						secondObserving = 0;
					async.parallel([
						function(callback){
							redisMutex1.getStatus(function(err, result){
								actualmuid = result.value;
								firstWaiting += result.waiting;
								firstObserving += result.observing;
								callback(null, true);
							});
						},
						function(callback){
							redisMutex2.getStatus(function(err, result){
								secondWaiting += result.waiting;
								secondObserving += result.observing;
								callback(null, true);
							});
						}
						],
						function(err, result){
							expect(actualmuid).toEqual(muid);
							expect(firstWaiting+secondWaiting).toEqual(1);
							expect(firstObserving+secondObserving).toEqual(1);
							done();
						}
					);
				}, 1500);

			}
		);

	}, 20000);

	it('release mutex', function(done){
		var redisMutex1 = redisSharedObject1.getMutexClient(testMutexKey1),
			redisMutex2 = redisSharedObject2.getMutexClient(testMutexKey1);		
		async.parallel([
			function(callback){
				redisMutex1.rel('wrong_id', function(err, result){
					if(err)
						console.log('err while releasing : ' + err);
					else
						console.log('released : ' + result);


					redisMutex1.rel(muid, function(err, result){
						if(err)
							console.log('err while releasing : ' + err);
						else
							console.log('released : ' + result);
						callback(err, result);
					});
				});						
			}
			],function(err, result){
				setTimeout( function(){ 
					var actualmuid,
						firstWaiting = 0,
						firstObserving = 0,
						secondWaiting = 0,
						secondObserving = 0;
					async.parallel([
						function(callback){
							redisMutex1.getStatus(function(err, result){
								actualmuid = result.value;
								firstWaiting += result.waiting;
								firstObserving += result.observing;
								callback(null, true);
							});
						},
						function(callback){
							redisMutex2.getStatus(function(err, result){
								secondWaiting += result.waiting;
								secondObserving += result.observing;
								callback(null, true);
							});
						}
						],
						function(err, result){
							expect(actualmuid).not.toBe(null);
							expect(firstWaiting+secondWaiting).toEqual(0);
							expect(firstObserving+secondObserving).toEqual(0);
							done();
						}
					);
				}, 2000);					
			}
		);

	}, 20000);


	it('this mutex will be expired', function(done){
		redisSharedObject1.createMutexClient('toBeExpired', 3, function(err, toBeExpiredSoon){
			toBeExpiredSoon.on('expired', function(expired_id){
				console.log(expired_id + ' has been expired');
				expect(muid).toEqual(expired_id);
			});	

			toBeExpiredSoon.get(function(err, result){
				if(err)
					console.log('err while waiting(5) : ' + err);
				else{
					console.log('got mutex(5) : ' + result);
					muid = result;
				}
				setTimeout(function(){
					redisSharedObject1.end();
					redisSharedObject2.end();
					done();
				}, 5000);
			});

		});
	}, 20000); 

});

test/j-runner.js
var Jasmine = require('jasmine');
var SpecReporter = require('jasmine-spec-reporter');
var noop = function() {};

var jrunner = new Jasmine();
jrunner.configureDefaultReporter({print: noop});    // remove default reporter logs
jasmine.getEnv().addReporter(new SpecReporter());   // add jasmine-spec-reporter
jrunner.loadConfigFile('test/jasmine.json');                           // load jasmine.json configuration
jrunner.execute();

lib/redisConnector.js
'use strict';

var Promise = require('bluebird'),
	redis = require('redis'),
	util = require('util'),
	EventEmitter = require('events');

if (typeof EventEmitter !== 'function') {
    EventEmitter = EventEmitter.EventEmitter;
}

Promise.promisifyAll(redis.RedisClient.prototype);

function RedisConnector(options, callback) {
	this.options = options;
	this.callback = callback;
	this._initialize(options);
};

module.exports = RedisConnector;

util.inherits(RedisConnector, EventEmitter);

var _ = RedisConnector.prototype;

_._connect = function(options){
	var that = this;
	that.client = redis.createClient(options.port, options.host, options);

	if(options.pass) {
		that.client.auth(options.pass, function(err){
			if(err && that.callback) {
				that.callback(err,null);
			}
		});
	}
	if(options.db)
		that.client.select(options.db);

	that.client.on('connect', function() {
		if(options.db)
			that.client.select(options.db);
		if(that.callback)
			that.callback(null, that.client);
	});

	that.client.on('error', function(err){	
		if(that.callback)
			that.callback(err,null);
	});
};

_._initialize = function(options) {
	var that = this;
    that._connect(options);
};

lib/factory.js
'use strict';

var Promise = require('bluebird'),
	redis = require('redis'),
	util = require('util'),
	SharedObject = require('./sharedObject'),
	RedisConnector = require('./redisConnector'),
	uuid = require('uuid');

exports.factory = function(connect_options, cb){
	var	mutexTimer = {},
		SharedObjectFactory = function(connect_options, cb) {		
			if(connect_options && connect_options.options){
				this.client = connect_options;
			}
			else{
				this.options = connect_options || {
					host: '127.0.0.1',
					port: 6379,
					db: 1
				};
			}
			this._initialize(cb);
		};

	function Semaphore(object, count, prefix, client, subscriber) {
		this.luaScripts = require('../script/semaphore');
		this.count = count;
		SharedObject.call(this, object, prefix, client, subscriber);
	}

	util.inherits(Semaphore, SharedObject);

	Semaphore.prototype._reset = function(cb){  
		var that = this;
		return new Promise(function(resolve, reject){
			that.client.set(that.key, that.count, function(err, result){
				if(err){
					reject(err);
				}
				else{
					if(result === "OK"){
						resolve(true);
					}
					else{
						reject(false);
					}
				}
			});
		});
	};

	Semaphore.prototype._get = function(){
		var that = this;

		return new Promise(function(resolve, reject){		
			that._evalJustInTime('get', 1, that.key, function(err, reply){
				if(err) {
					reject(err);
				}
				else if(reply === null){
					var keyNotFoundErr = new Error('key was deleted');
					keyNotFoundErr.code = 'ENOTFOUNDKEY';
					reject(keyNotFoundErr);
				}
				else if(reply >= 0) {
					resolve(true);	
					that._publish(that.prefixInfo + that.key, {
						name : 'semaphore_acquired',
						value : reply
					});	
				}
				else {	
					resolve(false);
				}
			});
		});

	};

	Semaphore.prototype._rel = function(){
		var that = this;

		return new Promise(function(resolve, reject){
			that.client.incr(that.key, function(err, result){
				if(err)
					reject(err);
				else{
					that._publish(that.key, {
						name : 'semaphore_released',
						count : result
					});
					resolve(result);		
				}
			});
		});
	};

	Semaphore.prototype.check = function(cb){
		return SharedObject.prototype.check.call(this, cb, function(reply){
			return reply > 0;
		});
	};

	function Mutex(object, prefix, client, subscriber, timeout){
		this.luaScripts = require('../script/mutex');
		this.timeout = timeout ? Math.floor(timeout * 1000) : 60000;
		SharedObject.call(this, object, prefix, client, subscriber);
	}

	util.inherits(Mutex, SharedObject);

	Mutex.prototype._reset = function(cb){
		var that = this;
		if(mutexTimer[this.key])
			clearTimeout(mutexTimer[this.key]);
		mutexTimer[this.key] = null;
		
		return new Promise(function(resolve, reject){
			that.client.del(that.key, function(err, result){
				if(err){
					reject(err);
				}
				else{
					if(result > 0){
						resolve(true);
					}
					else{
						resolve(false);
					}
				}
			});
		});
	};

	Mutex.prototype._get = function(){
		var that = this,
			muid = uuid.v4(),
			startTime = Date.now(),
			setMutexTimer = function(){
				that.timestamp = startTime;
				mutexTimer[that.key] = setTimeout(function(){
					mutexTimer[that.key] = null;
					that.emit('expired', 1);
					that._publish(that.key, {
						name : 'mutex_expired',
						count : 1
					});
				}, that.timeout);
				that._publish(that.prefixInfo + that.key, {
					name : 'mutex_locked',
					value : muid
				});
			};
		return new Promise(function(resolve, reject){		
			that._evalJustInTime('get', 1, that.key, muid, that.timeout, function(err, reply) {
				if(err){
					reject(err);
				}
				else if(reply === 1) {
					setMutexTimer();				
					resolve(muid);
				} else {
					resolve(null);
				}
			});
		});
	};

	Mutex.prototype._rel = function(muid){
		var that = this;
		return new Promise(function(resolve, reject){		
			that._evalJustInTime('rel', 1, that.key, muid, function(err, reply){
				if(err){
					reject(err);
				}
				else if(reply === -1){
					var noAccessError = new Error('Access denied');
					noAccessError.code = 'ENOACCESS';
					reject(noAccessError);				
				}
				else if(reply === null || mutexTimer[that.key] === null){
					resolve(false);
				}
				else { // if(muid === reply)
					if(reply > 0){
						clearTimeout(mutexTimer[that.key]);
						mutexTimer[that.key] = null;
						that._publish(that.key, {
							name : 'mutex_unlocked',
							count : 1
						});
					}
					resolve(reply > 0);
				}
			});
		});
	};

	Mutex.prototype.check = function(cb){
		return SharedObject.prototype.check.call(this, cb, function(reply){
			return !reply;
		});
	};

	Mutex.prototype._extend = function(muid, timeout){
		var that = this,
			newTimeout = that.timeout - (Date.now() - that.timestamp) + (timeout * 1000);
		return new Promise(function(resolve, reject){		
			that._evalJustInTime('extend', 1, that.key, muid, newTimeout, function(err, result){
				if(err){
					reject(err);
				}
				else{
					if(mutexTimer[that.key] === null){
						if(result){
							that.client.del(that.key, function(err, result){
								if(err)
									reject(err);
								else
									resolve(false);
							});
						}
						else
							resolve(false);
					}
					else if(result){
						clearTimeout(mutexTimer[that.key]);
						mutexTimer[that.key] = setTimeout(function(){
							that.emit('expired', 1);
							that._publish(that.key, {
								name : 'mutex_expired',
								count : 1
							});
						}, newTimeout);
						resolve(true);
					}
					else{
						resolve(false);
					}
				}
			});
		});
	};

	Mutex.prototype.extend = function(muid, timeout, cb){
		return this._convertToHybridCall(this._extend(muid, timeout), cb);
	};

	Mutex.prototype._debug_resetMutexTimer = function(){ // only for testing
		mutexTimer[this.key] = null;
	};

	var _ = SharedObjectFactory.prototype;

	_._initialize = function(cb){
		var that = this;
		that.prefixSemaphore = 'sema:';
		that.prefixMutex = 'mutex:';
		that.sharedObjects = {};

		if(that.client === undefined){		
			that.client = (new RedisConnector(that.options, function(err, redisClient){
				if(cb){
					if(err){
						return cb(err, null);
					}
					else{
						return cb(null, redisClient);
					}
				}
			})).client;
			that._internal_connection_used = true;	
			that.subscriber = redis.createClient(that.options.port, that.options.host, that.options);
			if(that.options.pass) {
				that.subscriber.auth(that.options.pass);
			}
		}
		else{
			that.subscriber = redis.createClient(that.client.options.port, that.client.options.host, that.client.options);
			if(that.client.options.pass) {
				that.subscriber.auth(that.client.options.pass);
			}
			if(cb)
				cb(null, that.client);
		}
	};

	_.end = function(){
		var key;
		for(key in mutexTimer){
			if(mutexTimer[key])
				clearTimeout(mutexTimer[key]);
		}
		mutexTimer  = {};
		for(key in this.sharedObjects){
			if({}.hasOwnProperty.call(this.sharedObjects, key))
				this.client.del(key);
		}
		this.subscriber.quit();
		if(this._internal_connection_used){
			this.client.quit();
		}
	};

	_._resetObject = function(key, cb){
		var that = this;
		if(cb && typeof cb === "function")
			that.sharedObjects[key].reset(function(err, result){
				if(err)
					cb(err, null);
				else
					cb(null, that.sharedObjects[key]);
			});	
		else{
			return that.sharedObjects[key].reset().then(function(result){
				return Promise.resolve(that.sharedObjects[key]);
			});
		}
	};

	_._createSemaphore = function(object, count){
		var that = this;
		that.sharedObjects[that.prefixSemaphore+object] = new Semaphore(object, count, that.prefixSemaphore, that.client, that.subscriber);
		that.sharedObjects[that.prefixSemaphore+object].setDefaultPriority(exports.priority.HIGH);
	};

	_.createSemaphoreClient = function(object, count, cb){
		var that = this;
		if(!{}.hasOwnProperty.call(that.sharedObjects, that.prefixSemaphore+object)){
			that._createSemaphore(object, count);
		}
		return that._resetObject(that.prefixSemaphore+object, cb);
	};

	_.getSemaphoreClient = function(object, count){
		var that = this;
		count = count === undefined ? 3 : count;
		if(!{}.hasOwnProperty.call(that.sharedObjects, that.prefixSemaphore+object)){
			that._createSemaphore(object, count);
		}
		return that.sharedObjects[that.prefixSemaphore+object];
	};

	_._createMutex = function(object, timeout){
		var that = this;
		that.sharedObjects[that.prefixMutex+object] = new Mutex(object, that.prefixMutex, that.client, that.subscriber, timeout);
		that.sharedObjects[that.prefixMutex+object].setDefaultPriority(exports.priority.HIGH);
	}

	_.createMutexClient = function(object, timeout, cb){
		var that = this;
		if(!{}.hasOwnProperty.call(that.sharedObjects, that.prefixMutex+object)){
			that._createMutex(object, timeout);
		}
		that.sharedObjects[that.prefixMutex+object].timeout = Math.floor(timeout * 1000);
		return that._resetObject(that.prefixMutex+object, cb);
	};

	_.getMutexClient = function(object, timeout){
		var that = this;
		timeout = timeout === undefined ? 10 : timeout;
		if(!{}.hasOwnProperty.call(that.sharedObjects, that.prefixMutex+object)){
			that._createMutex(object, timeout);
			that.sharedObjects[that.prefixMutex+object].timeout = Math.floor(timeout * 1000);
		}
		return that.sharedObjects[that.prefixMutex+object];
	};

	return new SharedObjectFactory(connect_options, cb);
};

exports.priority = {
		HIGH : 0,
		NORMAL : 15,
		LOW : 40
	};
